internal/packer_registry/types.bucket.go

```diff
@@ -0,0 +1,288 @@
+package packer_registry
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-packer-service/preview/2021-04-30/models"
+	"github.com/hashicorp/packer/internal/packer_registry/env"
+	"google.golang.org/grpc/codes"
+)
+
+type Bucket struct {
+	Slug        string
+	Description string
+	Destination string
+	Labels      map[string]string
+	Config      ClientConfig
+	*Iteration
+	client *Client
+}
+
+// NewBucketWithIteration initializes a simple Bucket that can be used for tracking Packer
+// related build bits.
+func NewBucketWithIteration(opts IterationOptions) (*Bucket, error) {
+
+	b := Bucket{
+		Labels: map[string]string{
+			"ManagedBy": "Packer",
+		},
+	}
+
+	i, err := NewIteration(opts)
+	if err != nil {
+		return nil, err
+	}
+	b.Iteration = i
+
+	return &b, nil
+}
+
+func (b *Bucket) Validate() error {
+	if b.Slug == "" {
+		return fmt.Errorf("no Packer bucket name defined; either the environment variable %q is undefined or the HCL configuration has no build name", env.HCPPackerBucket)
+	}
+
+	if b.Destination == "" {
+		return fmt.Errorf("no Packer registry defined; either the environment variable %q is undefined or the HCL configuration has no build name", env.HCPPackerRegistry)
+	}
+
+	return nil
+}
+
+// Initialize registers the Bucket b with the configured HCP Packer Registry.
+// Upon initialization a Bucket will be upserted to, a new iteration will be created for the build if the configured
+// fingerprint has not associated builds. Lastly, the initialization process with registered the builds that need to be
+// completed before an iteration can be marked as DONE.
+//
+// b.Initialize() must be called before any data can be published to the configured HCP Packer Registry.
+func (b *Bucket) Initialize(ctx context.Context) error {
+	// NOOP
+	if b == nil {
+		return nil
+	}
```

Comment:
NIT: It's probably just a matter of preference here but, I think that this is okay for this to panic, that probably means something went wrong in the programming side of things, if that Bucket is nil.
```suggestion
```

Reply:
Ditto throughout.

(also, if not the panic, then we should at least error)